### PR TITLE
feat(parameters): added `BaseProvider` class

### PR DESCRIPTION
**Pull request by dreamorosi**
## Description of your changes

This PR introduces a new package (`parameters`) in the npm workspace as well as the `BaseProvider` for the new utility and its unit tests.

The changes related to the new workspace package include:
- New `pacakges/parameters` folder with `package.json`
- TSConfig files
- Eslint file
- Jest test harness, and jestconfig file
- README file

The changes related to the `BaseProvider` include the content of the `src` and `tests` directories:
- New `BaseProvider` class
- New types and exceptions to support the class
- Unit tests for `BaseProvider`

> **Note**
> I have left a few comments/questions that I would appreciate reviewers to engage with during the review

Please don't take in account the content of the `README.md` file nor the docstrings (or lack thereof) for the various features, both items will be addressed in their own PRs.

### How to verify this change

See unit tests added as part of this PR.

### Related issues, RFCs

**Issue number:** #1173 

### PR status

***Is this ready for review?:*** YES  
***Is it a breaking change?:*** NO

## Checklist

- [x] [My changes meet the tenets criteria](https://awslabs.github.io/aws-lambda-powertools-python/#tenets)
- [x] I have performed a *self-review* of my own code
- [x] I have *commented* my code where necessary, particularly in areas that should be flagged with a TODO, or hard-to-understand areas
- [ ] I have made corresponding changes to the *documentation*
- [ ] I have made corresponding changes to the *examples*
- [x] My changes generate *no new warnings*
- [x] The *code coverage* hasn't decreased
- [x] I have *added tests* that prove my change is effective and works
- [x] New and existing *unit tests pass* locally and in Github Actions
- [ ] Any *dependent changes have been merged and published* in downstream module
- [x] The PR title follows the [conventional commit semantics](https://github.com/awslabs/aws-lambda-powertools-typescript/blob/main/.github/semantic.yml#L2)

### Breaking change checklist

- [ ] I have documented the migration process
- [ ] I have added, implemented necessary warnings (if it can live side by side)

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/parameters/src/BaseProvider.ts
Comment (dreamorosi) [inline packages/parameters/src/BaseProvider.ts:48]: Question: Should this be `public` or `private`?

File: packages/parameters/src/BaseProvider.ts
Comment (dreamorosi) [inline packages/parameters/src/BaseProvider.ts:104]: As mentioned in the comment here, I stayed generic with types because at this stage I'm not yet sure what each provider/AWS SDK (from `_get`) will return.

I have opened #1172 to track this work.

File: packages/parameters/src/BaseProvider.ts
Comment (dreamorosi) [inline packages/parameters/src/BaseProvider.ts:132]: I have opened #1172 to track this work.

File: packages/parameters/src/BaseProvider.ts
Comment (dreamorosi) [inline packages/parameters/src/BaseProvider.ts:170]: I have opened #1172 to track this work.

File: packages/parameters/src/BaseProvider.ts
Comment (dreamorosi) [inline packages/parameters/src/BaseProvider.ts:187]: I have opened #1172 to track this work.

File: packages/parameters/src/BaseProvider.ts
Comment (dreamorosi) [inline packages/parameters/src/BaseProvider.ts:157]: **As far as I can tell**, the Python implementation **doesn't** clear expired items from the cache.

Can you think of any case in which doing that might become an issue?

File: packages/parameters/src/BaseProvider.ts
Comment (ijemmy) [reply packages/parameters/src/BaseProvider.ts:48]: I'll go with `private` first and broaden later if needed.

File: packages/parameters/src/BaseProvider.ts
Comment (ijemmy) [inline packages/parameters/src/BaseProvider.ts:51]: (non-blocking) Just double check. This is redundant with the initialize in line 48. Is it intentional?

If it is (or it can complicate the type if we leave it possibly having `undefined` value), feel free to resolve this comment.

File: packages/parameters/src/BaseProvider.ts
Comment (ijemmy) [reply packages/parameters/src/BaseProvider.ts:20]: Given this file's size, I would say so. But it is not critical.

File: packages/parameters/src/BaseProvider.ts
Comment (ijemmy) [reply packages/parameters/src/BaseProvider.ts:80]: As long as the test cases are rigorous enough, this should be ok. I cannot see how we can run into having a duplicated here, can we?

File: packages/parameters/src/BaseProvider.ts
Comment (ijemmy) [reply packages/parameters/src/BaseProvider.ts:157]: No. We just trade time for space here. The operation is O(1) so I don't think there is any issue to delete it now. 

The problem is more on doing more than one thing in a single method. It is not obvious to me that `hasKeyExpiredInCache()` will prune the `this.store`.  The new maintainer may use this function and didn't know that he's pruned the cache. Thus, I'm incline to remove this.

If we want to prune expired items, I would expect a public/protected method to do this explicitly. But that will cause O(n) and I am not sure if it's useful.  Map has O(1) access so we are only saving the memory consumption here, which I expect it to be very small gain.

Thus, I'm incline to remove this and keep the expired item.

Comment (ijemmy) [reply packages/parameters/src/BaseProvider.ts:157]: That said. I don't think that it's bad enough to block this PR. So I would give an approval. You can decide in later commit if you want to remove this.

**Review by ijemmy — DISMISSED** [review]
One comment on prunning expire key

File: packages/parameters/src/BaseProvider.ts
Comment (dreamorosi) [reply packages/parameters/src/BaseProvider.ts:20]: Addressed in f9259eb9688525cbc26dfa0f4554369280d69504

File: packages/parameters/src/BaseProvider.ts
Comment (dreamorosi) [reply packages/parameters/src/BaseProvider.ts:48]: If I use `private` I'm not sure how to test around it. So I went with `private` and extended the class in the tests, similar to what done in the Idempotency utility.

If you have any better way to do it that also allows a pure `private` let me know.

File: packages/parameters/src/BaseProvider.ts
Comment (dreamorosi) [reply packages/parameters/src/BaseProvider.ts:80]: Agree.

I wasn't worried about duplicates as much as special characters and such.

File: packages/parameters/src/BaseProvider.ts
Comment (dreamorosi) [reply packages/parameters/src/BaseProvider.ts:157]: Good points, I didn't think about the operation costs and I agree that the memory savings are marginal.

I have removed it.

**Review by ijemmy — APPROVED** [review]

File: packages/parameters/src/BaseProvider.ts
Comment (ijemmy) [reply packages/parameters/src/BaseProvider.ts]: That should be good for now. I would expect that the tests cover all `privates` variable from the `public` methods.

File: packages/parameters/src/types/BaseProvider.ts
Comment (shdq) [inline packages/parameters/src/types/BaseProvider.ts:16]: Just to reduce boilerplate:
```
interface GetMultipleOptionsInterface extends GetOptionsInterface {
    throwOnTransformError?: boolean
}
```

Comment (dreamorosi) [reply packages/parameters/src/types/BaseProvider.ts:16]: Good point, please feel free to address this in your PR for `AppConfigprovider`

**Pull request merged**
